docsrv/docsrv.go

```diff
@@ -1,197 +1,142 @@
-	// or user, since there will be no need to use the mappings.yml file
-	// at all.
-	defaultOwner string
-	// baseFolder is the root folder served by the webserver.
-	baseFolder string
-	// sharedFolder is the location of the folder with all the shared assets.
-	sharedFolder string
-
-	github   GitHub
-	mappings mappings
-
-	mut *sync.RWMutex
-	// latestVersions contains a map from a ${owner}/${project} to a latest
-	// version, which is a version name with the time when it was last
-	// installed.
-	latestVersions map[string]latestVersion
-
-	installMut *sync.RWMutex
-	// installed is a set of installed versions in the format ${owner}/${project}/${version}.
-	installed map[string]struct{}
-
-	versionMut *sync.RWMutex
-	// versions contains a map of all versions available for a project.
-	versions map[string][]*version
+// Options contains all the options available for creating a new DocSrv
+// service.
+type Options struct {
+	// GitHubAPIKey is the API key used to retrieve releases from GitHub.
+	// If api key is empty the requests will be made without authentication.
+	GitHubAPIKey string
+	// DefaultOwner is the GitHub username or organization that will be used
+	// in case no mapping is found for a host. It's particularly useful if the
+	// documentations will always be for a specific organization, which would
+	// avoid requiring any mappings at all.
+	DefaultOwner string
+	// BaseFolder is the path to the root folder of the webserver.
+	BaseFolder string
+	// SharedFolder is the path to the folder used to store all the common
+	// assets for building the documentations.
+	SharedFolder string
+	// Mappings contains all the mappings between hosts and a repository
+	// in the form ${owner}/${project} (in GitHub).
+	Mappings Mappings
 }
 
-type version struct {
-	Text string `json:"text"`
-	URL  string `json:"url"`
+// Service is the main docsrv service.
+type Service struct {
+	opts    Options
+	fetcher releaseFetcher
+	index   *projectIndex
 }
 
-// latestVersion is a version name with the time it was inserted in the cache.
-type latestVersion struct {
-	cachedAt time.Time
-	version  string
-}
+// New creates a new DocSrv service with the given options.
+func New(opts Options) *Service {
+	if opts.Mappings == nil {
+		opts.Mappings = make(Mappings)
+	}
 
-// isExpired reports whether this version is expired or not and should be re-checked.
-func (v latestVersion) isExpired() bool {
-	return v.cachedAt.Add(latestVersionLifetime).After(time.Now())
+	return &Service{
+		opts:    opts,
+		fetcher: newReleaseFetcher(opts.GitHubAPIKey, 0),
+		index:   newProjectIndex(),
+	}
 }
 
-// latestVersionLifetime defines the time a latest version is valid.
-const latestVersionLifetime = 1 * time.Hour
+// ensureIndexed checks if the project is indexed and if it's not, it indexes
+// it.
+func (s *Service) ensureIndexed(owner, project string) error {
+	if !s.index.isIndexed(owner, project) {
```

Comment:
Then you lose the error you have to return at the end of the method.

Anyway, this method has changed in #15